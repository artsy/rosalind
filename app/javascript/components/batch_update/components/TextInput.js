import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

class TextInput extends React.Component {
  handleKeyUp = e => {
    if (e.key === 'Enter') {
      let value
      if (this.props.numeric) {
        value = parseInt(e.target.value)
        if (isNaN(value)) {
          return
        }
      } else {
        value = e.target.value
      }
      this.props.onEnter(value)
      e.target.value = null
    }
  }

  render() {
    const inputProps = {
      type: 'text',
      placeholder: this.props.placeholder,
      onKeyUp: this.handleKeyUp,
    }

    return (
      <div className={this.props.className}>
        <input {...inputProps} />
      </div>
    )
  }
}

TextInput.propTypes = {
  onEnter: PropTypes.func,
}

TextInput.defaultProps = {
  placeholder: 'Add some text',
}

const StyledTextInput = styled(TextInput)`
  input {
    width: 100%;
    margin: 1em 0 0 0;
    line-height: 2em;
    font-size: 1em;
    border: none;
    border-bottom: solid 1px #ddd;
    background: none;
  }

  input:focus {
    outline: none;
  }
`

export { StyledTextInput as default, TextInput }
