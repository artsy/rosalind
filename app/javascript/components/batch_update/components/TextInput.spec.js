import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { TextInput } from './TextInput'
import { mount } from 'enzyme'

it('renders correctly', () => {
  const rendered = renderer.create(<TextInput />)
  const tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()
})

it('submits the text upon pressing Enter', () => {
  const mockOnEnter = jest.fn()
  const wrapper = mount(<TextInput onEnter={mockOnEnter} />)

  wrapper.find('input').simulate('change', { target: { value: 'soup' } })
  wrapper.find('input').simulate('keyup', { key: 'Enter' })

  expect(mockOnEnter).toHaveBeenCalledTimes(1)
})

describe('numeric', () => {
  let component, onEnter, wrapper, input, tree
  beforeEach(() => {
    onEnter = jest.fn()
    component = <TextInput onEnter={onEnter} placeholder="foo" numeric />
    wrapper = mount(component)
    input = wrapper.find('input')
    tree = renderer.create(component).toJSON()
  })
  it('adds a keydown handler if a numeric prop is passed in', () => {
    expect(tree).toMatchSnapshot()
  })
  it('converts values to integers onEnter', () => {
    input.simulate('keyup', { key: 'Enter', target: { value: '123' } })
    expect(onEnter).toHaveBeenCalledWith(123)

    onEnter.mockClear()
    input.simulate('keyup', { key: 'Enter', target: { value: '123.9' } })
    expect(onEnter).toHaveBeenCalledWith(123)

    onEnter.mockClear()
    input.simulate('keyup', { key: 'Enter', target: { value: '123ABC' } })
    expect(onEnter).toHaveBeenCalledWith(123)
  })
  it('will not submit non-numeric values', () => {
    input.simulate('keyup', { key: 'Enter', target: { value: 'abc123' } })
    input.simulate('keyup', { key: 'Enter', target: { value: '$123' } })
    input.simulate('keyup', { key: 'Enter', target: { value: 'Four score' } })

    expect(onEnter).not.toHaveBeenCalled()
  })
})
