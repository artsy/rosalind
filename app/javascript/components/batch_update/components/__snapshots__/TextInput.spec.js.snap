// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`numeric adds a keydown handler if a numeric prop is passed in 1`] = `
<div
  className={undefined}
>
  <input
    onKeyUp={[Function]}
    placeholder="foo"
    type="text"
  />
</div>
`;

exports[`renders correctly 1`] = `
<div
  className={undefined}
>
  <input
    onKeyUp={[Function]}
    placeholder="Add some text"
    type="text"
  />
</div>
`;
